### ISSUE TITLE
Running Browserstack Local for Automate Tests

### ISSUE THREAD

**Issue by PaulHughes01** [Shippable/support#2885]
Hi there!

First want to say how happy I am with Shippable. We've got it fully integrated in our CI/CD flow and have zero complaints. :)

My question is related to integrating Shippable with Browserstack Automate. Currently, we run the automate tests after deploying to our actual test environment, but it would be better if we were able to run them in the container before deploy. Thankfully, Browserstack does provide [some binaries that can be run locally](https://www.browserstack.com/local-testing) to allow a tunnel for their tests to run on a server behind a firewall or in a VPN. It _does_ work when I build the docker container on my local dev machine, run the binary, and execute the tests by hand from within the container. However, when it runs within Shippable CI, it would appear that Browserstack is unable to connect (hence the "Page Not Found" error). Interestingly, the Binary does run with no errors.

Not sure if anyone else has experience getting Browserstack Automate working from within Shippable CI, but if so I would love some insight. Thanks in advance!
- Best example build: 57b73a6f1419e20f001a159f (there are some after this, but they're small "attempts" which can be ignored)

**Comment by rageshkrishna** [comment]
Hi @PaulHughes01. I did try BrowserStack a long time ago and was able to make it run successfully. I can't see anything that is obviously out of place in your build. If I remember correctly, sometimes BrowserStack take a couple of seconds to initialize fully. Your build, however, almost immediately proceeds to running the tests. My guess is that BrowserStackLocal isn't done starting up by then.

Could you try introducing a `sleep 10` (or more?) to add some delay before the tests? If that works, we'll know for sure that it's just BrowserStack's startup time that's causing this.

**Comment by PaulHughes01** [comment]
Hi, @rageshkrishna; thanks for the tip. Sadly, it doesn't seem to have done the trick (see build 57bb1f7a1419e20f001a98e1). :-/

I am glad to hear you did get it working at one point! One thing, we are running the tests inside our own container (being built pre_ci). Like I say, I *am* able to run locally using the same built container successfully, but thought it might be somehow related to that.

Out of curiosity, which image (if you remember) did you use previously when it worked for you? I'd be interested in trying to effectively build the same situation as you used before (not even necessarily using our application) to see if it sheds some light on it from my end. 

Thanks for all your help, it is much appreciated!

**Comment by rageshkrishna** [comment]
Hi @PaulHughes01. I checked, but it looks like I deleted that project. Let me try and whip up a new sample project and see how it goes.

**Comment by PaulHughes01** [comment]
Gotcha, thanks so much!

**Comment by PaulHughes01** [comment]
Hi, Ragesh—I hope all is well;

Wanted to check in on this and see if you had made any progress on the quick sample project. Thanks again!

**Comment by rageshkrishna** [comment]
Hi @PaulHughes01. I'm very sorry about the delay on this. I've been tied up with the recent release of our all new pipelines. I will put a sample together over the weekend and share it with you.

**Comment by PaulHughes01** [comment]
No problem; I expected you all have been busy with that release. Haven't dug in too deep to the new Pipelines but I'm looking forward to the opportunity to do so. :)

**Comment by rageshkrishna** [comment]
@PaulHughes01 It looks like that weekend took a lot longer than I had hoped :)

I've created a very basic sample that runs BrowserStackLocal on a nodejs project here: https://github.com/rageshkrishna/sample-browserstack-local

You can also see the successful Shippable build at https://app.shippable.com/runs/57d8fc76bb79950e00521120/1/console

**Comment by rageshkrishna** [comment]
In your case, it looks like BrowserStackLocal is working because I see that requests are being made to your app. Can you take a look at the video on BrowserStack and see what the browser shows for those requests when the build runs on Shippable?

**Comment by PaulHughes01** [comment]
Hey, thanks @rageshkrishna! I should be able to take a look at this later today or tomorrow and I'll let you know. 

Thanks again!

**Comment by rageshkrishna** [comment]
Hi @PaulHughes01. Just checking in to see if you've had any luck with this.

**Comment by PaulHughes01** [comment]
Thanks Ragesh!

I've tried to copy your code as much as is possible within our custom container (being built from a Dockerfile) and still am unable to get it to work within the Shippable CI run—but it does (still) work on my local machine from within the container. For that reason it _feels_ like there some reason it can't get into the container when running on Shippable. I see in your container it is listening on port 3000, and in mine it is exposing and listening on Port 80—would that be something that would be a block it?

**Comment by rageshkrishna** [comment]
@PaulHughes01 I'm sorry I lost track of this thread again for so long. Is this still happening?

**Comment by rageshkrishna** [comment]
Closing; feel free to reopen if this is still an issue. For what it's worth, I'm still unable to reproduce this with my projects.

**Comment by PaulHughes01** [comment]
Hi, Ragesh! 

Had to switch tasks off of this, and my colleague Jobaer ended up pulling back out this project recently. Anyway, we figured out where the problem was on our side, and it had to do with the fact that the Shippable builds override the entrypoint command specified in customer dockerfiles in the `docker run` command to launch the container. Our entrypoint called a script that launches Apache, so since that was not getting called, the actual Apache server was not booting. Of course our other phpunit and mocha tests did not catch that as an issue, as they could run without an actual Apache server running. 

So, we just had to add back in our launch script to the shippable.yml's `ci` section and then it worked perfectly. (In your node.js example, that happened by way of your npm start call).

Thanks for the help! I wanted to comment here even though it is closed and we got it solved, in case anyone else runs into the issue and the same brick wall. :)

Cheers!
